lib/engine/game/g_1858/step/track.rb

```diff
@@ -0,0 +1,183 @@
+            # lay as this can be paid by the company president.
+            action = get_tile_lay(entity)
+            return false unless action
+
+            # Check whether there are any hexes where track can be laid. After a
+            # couple of operating rounds many of the private railway companies
+            # will not be able to lay track, so this step (and so their entire
+            # operating turn) can be skipped if it not possible to add track.
+            hexes = (@game.graph_broad.connected_hexes(entity).keys +
+                     @game.graph_metre.connected_hexes(entity).keys).uniq
+            hexes.any? { |hex| available_hex(entity, hex) }
+          end
+
+          # If a public company does not have enough money to pay its terrain or
+          # extra tile costs then transfer money from the company's president to
+          # pay for these.
+          def subsidise_track(corporation, cost)
+            return unless corporation.corporation?
+            return unless cost.positive?
+            return unless cost > corporation.cash
+
+            president = corporation.owner
+            shortfall = cost - corporation.cash
+            if shortfall > corporation.owner.cash
+              raise GameError, "The tile lay costs #{@game.format_currency(cost)} but " \
+                               "#{corporation.name} has #{@game.format_currency(corporation.cash)} and " \
+                               "#{president.name} has #{@game.format_currency(president.cash)}"
+            end
+            @game.log << "#{president.name} contributes #{@game.format_currency(shortfall)} " \
+                         "towards the cost of #{corporation.name}'s tile lay"
+            president.spend(shortfall, corporation)
+          end
+
+          # Override the default implementation to give a better log message.
+          # This also allows a more intuitive name to be used for the method for
+          # the player contributing money for the tile, instead of abusing
+          # try_take_loan.
+          def pay_tile_cost!(entity, tile, rotation, hex, spender, cost, _extra_cost)
+            subsidise_track(spender, cost)
+            spender.spend(cost, @game.bank) if cost.positive?
+
+            message = ''
+            message += "#{spender.name} spends #{@game.format_currency(cost)} and " if cost.positive?
+            message += "#{entity.name} " if cost.zero? || (entity != spender)
+            message += "lays tile ##{tile.name}"
+            message += " with rotation #{rotation} on #{hex.name}"
+            message += " (#{tile.location_name})" unless tile.location_name.to_s.empty?
+            @log << message
+          end
+
+          def hex_neighbors(entity, hex)
+            hexes_broad = @game.graph_broad.connected_hexes(entity)[hex]
+            hexes_metre = @game.graph_metre.connected_hexes(entity)[hex]
+            hexes = [hexes_broad, hexes_metre].compact.inject([], :|)
```

Comment:
Does this work?
```suggestion
            hexes = [*hexes_broad, *hexes_metre].uniq
```

Reply:
Yes, that works and is neater. I hadn't realised that you could use the splat operator on a nil value inside an array definition.